shuffle-plugin/src/main/scala/com/nvidia/spark/rapids/shuffle/ucx/UCX.scala

```diff
@@ -458,31 +475,68 @@ class UCX(executorId: Int, usingWakeupFeature: Boolean = true) extends AutoClose
     }
   }
 
-  def register(rootBuffer: MemoryBuffer, mmapCallback: MemoryRegistrationCallback): Unit = {
-    onWorkerThreadAsync(() => {
-      val mmapParam = new UcpMemMapParams()
-        .setAddress(rootBuffer.getAddress)
-        .setLength(rootBuffer.getLength)
-
-      //note that this can throw, lets call back and let caller figure out how to handle
-      try {
-        registeredMemory += context.memoryMap(mmapParam)
-        mmapCallback(true)
-      } catch {
-        case t: Throwable =>
-          logError(s"There was an issue registering $rootBuffer against UCX", t)
-          mmapCallback(false)
-      }
-    })
+  /**
+   * Return rkeys (if we have registered memory)
+   */
+  private lazy val localRkeys: Seq[ByteBuffer] = registeredMemory.synchronized {
+    while (pendingRegistration) {
+      registeredMemory.wait(100)
+    }
+    registeredMemory.map(_.getRemoteKeyBuffer)
   }
 
+  /**
+   * Register a set of `MemoryBuffers` against UCX.
+   *
+   * @param buffers to register
+   * @param mmapCallback callback invoked when the memory map operation completes or fails
+   */
+  def register(buffers: Seq[MemoryBuffer], mmapCallback: MemoryRegistrationCallback): Unit =
+    registeredMemory.synchronized {
+      pendingRegistration = true
+
+      onWorkerThreadAsync(() => {
+        var error: Throwable = null
+        try {
+          buffers.foreach { buffer =>
+            val mmapParam = new UcpMemMapParams()
+                .setAddress(buffer.getAddress)
+                .setLength(buffer.getLength)
+
+            //note that this can throw, lets call back and let caller figure out how to handle
+            try {
+              val registered = context.memoryMap(mmapParam)
+              registeredMemory.synchronized {
```

Comment:
Does it make sense to lock over and over here?  `pendingRegistration` is true, so there shouldn't be anyone waiting  on this lock, instead they'll be waiting to see `pendingRegistration` turn true.  Seems like we should grab the lock outside of `buffers.foreach` to reduce the locking overhead without any downsides of slowing another thread down trying to access.  Thoughts?

Reply:
Yeap that's better for sure, fixing.